doc: Add documentation for backup feature (#1547)
devfile/devworkspace-operator: docs/dwo-configuration.md

The backup makes a snapshot of Workspace PVCs and stores them as tar.gz archives in the specified OCI registry.
**Note:** By default, the DevWorkspace backup job is disabled.

Backup CronJob configuration fields:

- **`enable`**: Set to `true` to enable the backup job, `false` to disable it. Default: `false`.
- **`schedule`**: A Cron expression defining how often the backup job runs. Default: `"0 1 * * *"`.
- **`registry.path`**: A base registry location where the backup archives will be pushed.
The value provided for registry.path is only the first segment of the final location. The full registry path is assembled dynamically, incorporating the name of the workspace and the :latest tag, following this pattern:
`<registry.path>/<devworkspace-name>:latest`

- **`registry.authSecret`**: (Optional) The name of the Kubernetes Secret containing credentials to access the OCI registry. If not provided, it is assumed that the registry is public or uses integrated OpenShift registry.
- **`oras.extraArgs`**: (Optional) Additional arguments to pass to the `oras` CLI tool during push and pull operations.

There are several configuration options to customize the logic:

> ### Integrated OpenShift container registry

Reviewer: ```suggestion
### Integrated OpenShift container registry
```
Actually, disregard my previous comment, since it looks like this is an official naming - https://docs.redhat.com/en/documentation/openshift_container_platform/4.20/html/registry/registry-overview#openshift-registry-common-terms_registry-overview

This option is available only on OpenShift clusters with integrated container registry enabled and requires no additional configuration.

To enable the backup use following configuration in the global DWOC:

```yaml
apiVersion: controller.devfile.io/v1alpha1

> kind: DevWorkspaceOperatorConfig

Author: Fixed

config:
  routing:
    defaultRoutingClass: basic
  workspace:
    backupCronJob:
      enable: true
      registry:
        path: default-route-openshift-image-registry.apps.{cluster ID}.openshiftapps.com
      schedule: '0 */4 * * *' # cron expression with backup frequency
    imagePullPolicy: Always
```

**Note:** The `path` field must contain the URL to your OpenShift integrated registry given by the cluster.

Once the backup job is finished the backup archives will be available in the DevWorkspace namespace under a repository
with matching Devworkspace name.

### Regular OCI compatible registry
To use a regular OCI compatible registry for backups, you need to provide registry credentials. Depending on your
RBAC policy the token can be provided via a secret in the operator namespace or in each DevWorkspace namespace.
Having the secret in the DevWorkspace namespace allows for using different registry accounts per namespace with more
granular access control.

```yaml
kind: DevWorkspaceOperatorConfig

> apiVersion: controller.devfile.io/v1alpha1

Reviewer 2: ```suggestion
apiVersion: controller.devfile.io/v1alpha1
metadata:
  name: devworkspace-operator-config
  namespace: $OPERATOR_INSTALL_NAMESPACE
```

config:
  routing:
    defaultRoutingClass: basic
  workspace:
    backupCronJob:
      enable: true
      registry:
        authSecret: my-secret
        path: quay.io/my-company-org
      schedule: '0 */4 * * *'
    imagePullPolicy: Always
```

> The `authSecret` must point to real Kubernetes Secret of type `kubernetes.io/dockerconfigjson` containing credentials to access the registry.

Author: Fixed

- **`retainTime`**: The duration time in seconds since a DevWorkspace was last started before it is considered stale and eligible for cleanup. Default: 2592000 seconds (30 days).
- **`dryRun`**: Set to `true` to run the cleanup job in dry-run mode. In this mode, the job logs which DevWorkspaces would be removed but does not actually delete them. Set to `false` to perform the actual deletion. Default: `false`.

## Configuring Backup CronJob

The DevWorkspace backup job allows for periodic backups of DevWorkspace data to a specified backup location.
Once enabled and configured, the backup job will run at defined intervals to create backups of DevWorkspace data.
The backup controller depends on OCI compatible registry that is used as an artifact storage for backup archives.

> The backup makes a snapshoft of Workspace PVCs and stores them as tar.gz archives in the specified OCI registry.

Author: fixed

To enable the backup use following configuration in the global DWOC:

```yaml
apiVersion: controller.devfile.io/v1alpha1
kind: DevWorkspaceOperatorConfig
config:
  routing:
    defaultRoutingClass: basic
  workspace:
    backupCronJob:
      enable: true
      registry:
        path: default-route-openshift-image-registry.apps.{cluster ID}.openshiftapps.com
      schedule: '0 */4 * * *' # cron expression with backup frequency
    imagePullPolicy: Always
```

**Note:** The `path` field must contain the URL to your OpenShift integrated registry given by the cluster.

Once the backup job is finished the backup archives will be available in the DevWorkspace namespace under a repository
with matching Devworkspace name.

### Regular OCI compatible registry
To use a regular OCI compatible registry for backups, you need to provide registry credentials. Depending on your
RBAC policy the token can be provided via a secret in the operator namespace or in each DevWorkspace namespace.
Having the secret in the DevWorkspace namespace allows for using different registry accounts per namespace with more

> granular access controll.

Author: Fixed

- **`retainTime`**: The duration time in seconds since a DevWorkspace was last started before it is considered stale and eligible for cleanup. Default: 2592000 seconds (30 days).
- **`dryRun`**: Set to `true` to run the cleanup job in dry-run mode. In this mode, the job logs which DevWorkspaces would be removed but does not actually delete them. Set to `false` to perform the actual deletion. Default: `false`.

## Configuring Backup CronJob

The DevWorkspace backup job allows for periodic backups of DevWorkspace data to a specified backup location.
Once enabled and configured, the backup job will run at defined intervals to create backups of DevWorkspace data.

> The backup controller depends on OCI compatible registry that is used as an artifact storage for backup archives.

Author: Fixed

The backup makes a snapshot of Workspace PVCs and stores them as tar.gz archives in the specified OCI registry.
**Note:** By default, the DevWorkspace backup job is disabled.

Backup CronJob configuration fields:

- **`enable`**: Set to `true` to enable the backup job, `false` to disable it. Default: `false`.
- **`schedule`**: A Cron expression defining how often the backup job runs. Default: `"0 1 * * *"`.

> - **`registry.path`**: A registry location where the backup archives will be pushed. The full path is given by `<registry.path>/<devworkspace-name>:latest`.

Reviewer 3: Okay, maybe we can rephrase it to be a bit clearer:
```suggestion
- **`registry.path`**: A registry location where the backup archives will be pushed. The full image path is assembled by the controller using this base path, along with the dev workspace name and a tag. For example:`<registry.path>/<devworkspace-name>:latest`.
```
